neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,182 @@
+
+## Specification
+
+### Background
+
+We will use the fungible tokens (FT) contract as an example contract to explain the specification.  This section will briefly explain how this contract works on the network today.
+
+The contract consists of state where all the user' account balances are stored in a single HashMap data structure.  When a user wishes to transfer some FT to another user, the following steps take place:
+
+- The sender sends a transaction with a function call action to the contract to transfer the tokens from one account to another.
+- This transaction is routed to the shard where the user's account lives where it is converted to a receipt.
+- Then the receipt is routed to the shard where the FT contract lives.
+- Once it arrives on the FT contract's shard, the receipt is executed, the function call is performed, and the transfer is performed.
+
+In the lucky case, the sender's account resides on the same shard as the FT contract.  In this case, depending on the load on the network, the transaction may be converted into a receipt and the receipt is executed in the same block.  Assuming a uniform distribution of user accounts across the network, such cases will only happen for every `N`th transaction where `N` is the number of shards on the network.  In other cases, the minimum latency is 2 blocks.  In the first block, the transaction is converted into a receipt and transferred to the receiver's shard.  And in the second block, the receipt is processed.
+
+Additionally, note that each FT transfer requires exactly one function call.
+
+### Limitations
+
+Since all the account balances are stored in a single contract, this one contract has to be invoked in order to make any transfers.  Therefore, the maximum TPS of this contract is the maximum TPS of the shard it is deployed on.  The only way to increase this capacity would be to increase the capacity of the shard.  Adding more shards does not help.
+
+### Sharded FT contract
+
+In the centralised FT contract above, all the state is stored in a single centralised location.  The opposite extreme of this approach would be to store all the state in as distributed a manner as possible i.e. the account balances of each user is stored locally on their accounts instead.
+
+If the state were stored in such a distributed manner, performing a FT transfer would require the following steps.
+
+- The sender would send a transaction to their account to transfer some tokens to the receiver.
+- The transaction is converted to a receipt and the receipt is guaranteed to execute on the same shard as the account to execute it against is the same account that is used to convert the transaction into a receipt.
+- When the receipt is executed, the sender's balance is deducted and a cross contract receipt is sent to the receiver's account to increase its balance.
+- The receipt arrives at the receiver's shard and is executed, which increases the receiver's account balance.
```

Comment:
This is still somewhat vague. Please be more precise with sender being `alice.near` and receiver being `bob.near`. What exactly is the function call and where are the balances stored?